src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.BatchingDocumentCollection.cs

```diff
@@ -439,55 +452,66 @@ await _project._projectSystemProjectFactory.ApplyBatchChangeToWorkspaceAsync((so
         /// <summary>
         /// Process file content changes
         /// </summary>
-        /// <param name="projectSystemFilePath">filepath given from project system</param>
-        /// <param name="workspaceFilePath">filepath used in workspace. it might be different than projectSystemFilePath</param>
+        /// <param name="projectSystemFilePath">File path given from project system for the .cshtml file</param>
+        /// <param name="workspaceFilePath">File path for the equivalent .cs document used in workspace. it might be different than projectSystemFilePath.</param>
         public void ProcessDynamicFileChange(string projectSystemFilePath, string workspaceFilePath)
         {
-            using (_project._gate.DisposableWait())
+            InterlockedOperations.Initialize(ref _dynamicFilesToRefresh, () =>
             {
-                // If our project has already been removed, this is a stale notification, and we can disregard.
-                if (_project.HasBeenRemoved)
-                {
-                    return;
-                }
+                return new AsyncBatchingWorkQueue<(string, string)>(
+                    TimeSpan.FromMilliseconds(200), // 200 chosen with absolutely no evidence whatsoever
+                    ProcessDynamicFileChangesAsync,
+                    EqualityComparer<(string, string)>.Default, // uses ordinal string comparison which is what we want
+                    _project._projectSystemProjectFactory.WorkspaceListener,
+                    _project._asynchronousFileChangeProcessingCancellationTokenSource.Token);
+            });
+
+            _dynamicFilesToRefresh.AddWork((projectSystemFilePath, workspaceFilePath));
+        }
+
+        private async ValueTask ProcessDynamicFileChangesAsync(ImmutableSegmentedList<(string projectSystemFilePath, string workspaceFilePath)> batch, CancellationToken cancellationToken)
+        {
+            foreach (var (projectSystemPath, workspaceFilePath) in batch)
+            {
+                DocumentId? documentId;
+                IDynamicFileInfoProvider? fileInfoProvider;
 
-                if (_documentPathsToDocumentIds.TryGetValue(workspaceFilePath, out var documentId))
+                using (await _project._gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
>DisposableWaitAsync [](http://example.com/codeflow?start=44&length=19)

Instead of waiting on this gate inside the loop, could we wait on it once and do a loop inside that to collect the items which need to be processed? A subsequent loop over the items to process could then do the GetDynamicFileInfoAsync/ApplyChangeToWorkspaceAsync calls